doc/source/configuration/settings.rst
.. _install-settings:

==================
Settings Reference
==================

Introduction
============

Horizon's settings broadly fall into four categories:

* `General Settings`_: this includes visual settings like the modal backdrop
  style, bug url and theme configuration, as well as settings that affect every
  service, such as page sizes on API requests.
* `Service-specific Settings`_: Many services that Horizon consumes, such
  as Nova and Neutron, don't advertise their capabilities via APIs, so Horizon
  carries configuration for operators to enable or disable many items. This
  section covers all settings that are specific to a single service.
* `Django Settings`_, which are common to all Django applications. The only
  ones documented here are those that Horizon alters by default; however, you
  should read the `Django settings documentation
  <https://docs.djangoproject.com/en/dev/topics/settings/>`_ to see the other
  options available to you.
* `Other Settings`_: settings which do not fall into any of the above
  categories.

To modify your settings, you have two options:

* **Preferred:** Add ``.py`` settings snippets to the
  ``openstack_dashboard/local/local_settings.d/`` directory. Several example
  files (appended with ``.example``) can be found there. These must start
  with an underscore, and are evaluated alphabetically, after
  ``local_settings.py``.
* Modify your ``openstack_dashboard/local/local_settings.py``. There is an
  file found at ``openstack_dashboard/local/local_settings.py.example``.

General Settings
================

.. _angular_features:

ANGULAR_FEATURES
----------------

.. versionadded:: 10.0.0(Newton)


.. deprecated:: 22.2.0(Zed)
   The Django version of the Images, Keypairs, and Roles panels are deprecated.
   As a result, "images_panel", "key_pairs_panel" and "roles_panel" keys in
   this setting are deprecated. Consider switching to the AngujarJS version
   by setting corresponding entries to True (the default value).

Default:

.. code-block:: python

    {
        'images_panel': True,
        'key_pairs_panel': True,
        'flavors_panel': False,
        'domains_panel': False,
        'users_panel': False,
        'groups_panel': False,
        'roles_panel': True
    }

A dictionary of currently available AngularJS features. This allows simple
toggling of legacy or rewritten features, such as new panels, workflows etc.

.. note::

    If you toggle ``domains_panel`` to ``True``, you also need to enable the
    setting of `OPENSTACK_KEYSTONE_DEFAULT_DOMAIN`_ and add
    `OPENSTACK_KEYSTONE_DEFAULT_DOMAIN`_ to `REST_API_REQUIRED_SETTINGS`_.

API_RESULT_LIMIT
----------------

.. versionadded:: 2012.1(Essex)

Default: ``1000``

The maximum number of objects (e.g. Swift objects or Glance images) to display
on a single page before providing a paging element (a "more" link) to paginate
results.

API_RESULT_PAGE_SIZE
--------------------

.. versionadded:: 2012.2(Folsom)

Default: ``20``

Similar to ``API_RESULT_LIMIT``. This setting controls the number of items
to be shown per page if API pagination support for this exists.

.. _available_themes:

AVAILABLE_THEMES
----------------

.. versionadded:: 9.0.0(Mitaka)

Default:

.. code-block:: python

   AVAILABLE_THEMES = [
        ('default', 'Default', 'themes/default'),
        ('material', 'Material', 'themes/material'),
   ]

This setting tells Horizon which themes to use.

A list of tuples which define multiple themes. The tuple format is
``('{{ theme_name }}', '{{ theme_label }}', '{{ theme_path }}')``.

The ``theme_name`` is the name used to define the directory which
the theme is collected into, under ``/{{ THEME_COLLECTION_DIR }}``.
It also specifies the key by which the selected theme is stored in
the browser's cookie.

The ``theme_label`` is the user-facing label that is shown in the
theme picker.  The theme picker is only visible if more than one
theme is configured, and shows under the topnav's user menu.

By default, the ``theme path`` is the directory that will serve as
the static root of the theme and the entire contents of the directory
is served up at ``/{{ THEME_COLLECTION_DIR }}/{{ theme_name }}``.
If you wish to include content other than static files in a theme
directory, but do not wish that content to be served up, then you
can create a sub directory named ``static``. If the theme folder
contains a sub-directory with the name ``static``, then
``static/custom/static`` will be used as the root for the content
served at ``/static/custom``.

The static root of the theme folder must always contain a _variables.scss
file and a _styles.scss file.  These must contain or import all the
bootstrap and horizon specific variables and styles which are used to style
the GUI. For example themes, see: /horizon/openstack_dashboard/themes/

Horizon ships with two themes configured. 'default' is the default theme,
and 'material' is based on Google's Material Design.

DEFAULT_POLICY_FILES
--------------------

.. versionadded:: 19.1.0(Wallaby)

Default:

.. code-block:: python

    {
        'identity': 'default_policies/keystone.yaml',
        'compute': 'default_policies/nova.yaml',
        'volume': 'default_policies/cinder.yaml',
        'image': 'default_policies/glance.yaml',
        'network': 'default_policies/neutron.yaml',
    }

This is a mapping from service types to YAML files including default
policy definitions. Values of this mapping should be relative paths to
`POLICY_FILES_PATH`_ or absolute paths. Policy files specified in this
setting are generated from default policies of back-end services,
so you rarely need to configure it. If you would like to override the
default policies, consider customizing files under `POLICY_FILES`_.

DEFAULT_THEME
-------------

.. versionadded:: 9.0.0(Mitaka)

Default: ``"default"``

This setting tells Horizon which theme to use if the user has not
yet selected a theme through the theme picker and therefore set the
cookie value. This value represents the ``theme_name`` key that is
used from `AVAILABLE_THEMES`_.  To use this setting, the theme must
also be configured inside of ``AVAILABLE_THEMES``. Your default theme
must be configured as part of `SELECTABLE_THEMES`_.  If it is not, then
your ``DEFAULT_THEME`` will default to the first theme in
``SELECTABLE_THEMES``.

DISALLOW_IFRAME_EMBED
---------------------

.. versionadded:: 8.0.0(Liberty)

Default: ``True``

This setting can be used to defend against Clickjacking and prevent Horizon from
being embedded within an iframe. Legacy browsers are still vulnerable to a
Cross-Frame Scripting (XFS) vulnerability, so this option allows extra security
hardening where iframes are not used in deployment. When set to true, a
``"frame-buster"`` script is inserted into the template header that prevents the
web page from being framed and therefore defends against clickjacking.

For more information see: http://tinyurl.com/anticlickjack

.. note::

  If your deployment requires the use of iframes, you can set this setting to
  ``False`` to exclude the frame-busting code and allow iframe embedding.

DROPDOWN_MAX_ITEMS
------------------

.. versionadded:: 2015.1(Kilo)

Default: ``30``

This setting sets the maximum number of items displayed in a dropdown.
Dropdowns that limit based on this value need to support a way to observe
the entire list.

FILTER_DATA_FIRST
-----------------

.. versionadded:: 10.0.0(Newton)

Default:

.. code-block:: python

    {
        'admin.instances': False,
        'admin.images': False,
        'admin.networks': False,
        'admin.routers': False,
        'admin.volumes': False
    }

If the dict key-value is True, when the view loads, an empty table will be
rendered and the user will be asked to provide a search criteria first (in case
no search criteria was provided) before loading any data.

Examples:

Override the dict:

.. code-block:: python

    {
        'admin.instances': True,
        'admin.images': True,
        'admin.networks': False,
        'admin.routers': False,
        'admin.volumes': False
    }

Or, if you want to turn this on for an specific panel/view do:

.. code-block:: python

    FILTER_DATA_FIRST['admin.instances'] = True

HORIZON_CONFIG
--------------

A dictionary of some Horizon configuration values. These are primarily
separated for historic design reasons.

Default:

.. code-block:: python

    HORIZON_CONFIG = {
        'user_home': 'openstack_dashboard.views.get_user_home',
        'ajax_queue_limit': 10,
        'auto_fade_alerts': {
            'delay': 3000,
            'fade_duration': 1500,
            'types': [
                'alert-success',
                'alert-info'
            ]
        },
        'bug_url': None,
        'help_url': "https://docs.openstack.org/",
        'exceptions': {
            'recoverable': exceptions.RECOVERABLE,
            'not_found': exceptions.NOT_FOUND,
            'unauthorized': exceptions.UNAUTHORIZED
        },
        'modal_backdrop': 'static',
        'angular_modules': [],
        'js_files': [],
        'js_spec_files': [],
        'external_templates': [],
    }

ajax_poll_interval
~~~~~~~~~~~~~~~~~~

.. versionadded:: 2012.1(Essex)

Default: ``2500``

How frequently resources in transition states should be polled for updates,
expressed in milliseconds.

ajax_queue_limit
~~~~~~~~~~~~~~~~

.. versionadded:: 2012.1(Essex)

Default: ``10``

The maximum number of simultaneous AJAX connections the dashboard may try
to make. This is particularly relevant when monitoring a large number of
instances, volumes, etc. which are all actively trying to update/change state.

angular_modules
~~~~~~~~~~~~~~~

.. versionadded:: 2014.2(Juno)

Default: ``[]``

A list of AngularJS modules to be loaded when Angular bootstraps. These modules
are added as dependencies on the root Horizon application ``horizon``.

auto_fade_alerts
~~~~~~~~~~~~~~~~

.. versionadded:: 2013.2(Havana)

Default:

.. code-block:: python

    {
        'delay': 3000,
        'fade_duration': 1500,
        'types': []
    }

If provided, will auto-fade the alert types specified. Valid alert types
include: ['alert-default', 'alert-success', 'alert-info', 'alert-warning',
'alert-danger']  Can also define the delay before the alert fades and the fade
out duration.

bug_url
~~~~~~~

.. versionadded:: 9.0.0(Mitaka)

Default: ``None``

If provided, a "Report Bug" link will be displayed in the site header which
links to the value of this setting (ideally a URL containing information on
how to report issues).

disable_password_reveal
~~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 2015.1(Kilo)

Default: ``False``

Setting this to True will disable the reveal button for password fields,
including on the login form.

exceptions
~~~~~~~~~~

.. versionadded:: 2012.1(Essex)

Default:

.. code-block:: python

    {
        'unauthorized': [],
        'not_found': [],
        'recoverable': []
    }

A dictionary containing classes of exceptions which Horizon's centralized
exception handling should be aware of. Based on these exception categories,
Horizon will handle the exception and display a message to the user.

help_url
~~~~~~~~

.. versionadded:: 2012.2(Folsom)

Default: ``None``

If provided, a "Help" link will be displayed in the site header which links
to the value of this setting (ideally a URL containing help information).

js_files
~~~~~~~~

.. versionadded:: 2014.2(Juno)

Default: ``[]``

A list of javascript source files to be included in the compressed set of files
that are loaded on every page. This is needed for AngularJS modules that are
referenced in ``angular_modules`` and therefore need to be include in every
page.

js_spec_files
~~~~~~~~~~~~~

.. versionadded:: 2015.1(Kilo)

Default: ``[]``

A list of javascript spec files to include for integration with the Jasmine
spec runner. Jasmine is a behavior-driven development framework for testing
JavaScript code.

modal_backdrop
~~~~~~~~~~~~~~

.. versionadded:: 2014.2(Kilo)

Default: ``"static"``

Controls how bootstrap backdrop element outside of modals looks and feels.
Valid values are ``"true"`` (show backdrop element outside the modal, close
the modal after clicking on backdrop), ``"false"`` (do not show backdrop
element, do not close the modal after clicking outside of it) and ``"static"``
(show backdrop element outside the modal, do not close the modal after
clicking on backdrop).

password_autocomplete
~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 2013.1(Grizzly)

Default: ``"off"``

Controls whether browser autocompletion should be enabled on the login form.
Valid values are ``"on"`` and ``"off"``.

password_validator
~~~~~~~~~~~~~~~~~~

.. versionadded:: 2012.1(Essex)

Default:

.. code-block:: python

    {
        'regex': '.*',
        'help_text': _("Password is not accepted")
    }

A dictionary containing a regular expression which will be used for password
validation and help text which will be displayed if the password does not
pass validation. The help text should describe the password requirements if
there are any.

This setting allows you to set rules for passwords if your organization
requires them.

user_home
~~~~~~~~~

.. versionadded:: 2012.1(Essex)

Default: ``settings.LOGIN_REDIRECT_URL``

This can be either a literal URL path (such as the default), or Python's
dotted string notation representing a function which will evaluate what URL
a user should be redirected to based on the attributes of that user.

MESSAGES_PATH
-------------

.. versionadded:: 9.0.0(Mitaka)

Default: ``None``

The absolute path to the directory where message files are collected.

When the user logins to horizon, the message files collected are processed
and displayed to the user. Each message file should contain a JSON formatted
data and must have a .json file extension. For example:

.. code-block:: python

    {
        "level": "info",
        "message": "message of the day here"
    }

Possible values for level are: ``success``, ``info``, ``warning`` and
``error``.

NG_TEMPLATE_CACHE_AGE
---------------------

.. versionadded:: 10.0.0(Newton)

Angular Templates are cached using this duration (in seconds) if `DEBUG`_
is set to ``False``.  Default value is ``2592000`` (or 30 days).

OPENSTACK_API_VERSIONS
----------------------

.. versionadded:: 2013.2(Havana)

Default:

.. code-block:: python

    {
        "identity": 3,
        "volume": 3,
        "compute": 2
    }

Overrides for OpenStack API versions. Use this setting to force the
OpenStack dashboard to use a specific API version for a given service API.

.. note::

    The version should be formatted as it appears in the URL for the
    service API. For example, the identity service APIs have inconsistent
    use of the decimal point, so valid options would be "3".
    For example:

    .. code-block:: python

        OPENSTACK_API_VERSIONS = {
            "identity": 3,
            "volume": 3,
            "compute": 2
        }

OPENSTACK_CLOUDS_YAML_CUSTOM_TEMPLATE
-------------------------------------

.. versionadded:: 15.0.0(Stein)

Default: ``None``

Example:: ``my-clouds.yaml.template``

A template name for a custom user's ``clouds.yaml`` file.
``None`` means the default template for ``clouds.yaml`` is used.

If the default template is not suitable for your deployment,
you can provide your own clouds.yaml by specifying this setting.

The default template is defined as :opendev-repo:`clouds.yaml.template
<openstack/horizon/src/branch/master/openstack_dashboard/dashboards/project/api_access/templates/api_access/clouds.yaml.template>`
and available context parameters are found in ``_get_openrc_credentials()``
and ``download_clouds_yaml_file()`` functions in
:opendev-repo:`openstack_dashboard/dashboards/project/api_access/views.py
<openstack/horizon/src/branch/master/openstack_dashboard/dashboards/project/api_access/views.py>`

.. note::

   Your template needs to be placed in the search paths of Django templates.
   You may need to configure `ADD_TEMPLATE_DIRS`_ setting
   to contain a path where your template exists.

OPENSTACK_CLOUDS_YAML_NAME
--------------------------

.. versionadded:: 12.0.0(Pike)

Default: ``"openstack"``

The name of the entry to put into the user's clouds.yaml file.

OPENSTACK_CLOUDS_YAML_PROFILE
-----------------------------

.. versionadded:: 12.0.0(Pike)

Default: ``None``

If set, the name of the `vendor profile`_ from `os-client-config`_.

.. _vendor profile: https://docs.openstack.org/os-client-config/latest/user/vendor-support.html
.. _os-client-config: https://docs.openstack.org/os-client-config/latest/

OPENSTACK_ENDPOINT_TYPE
-----------------------

.. versionadded:: 2012.1(Essex)

Default: ``"publicURL"``

A string which specifies the endpoint type to use for the endpoints in the
Keystone service catalog.

OPENSTACK_KEYSTONE_ENDPOINT_TYPE
--------------------------------

.. versionadded:: 23.1.0(Antelope)

Default: ``"publicURL"``

A string which specifies the endpoint type to use for the Keystone (identity)
endpoint when looking it up in the service catalog.

OPENSTACK_HOST
--------------

.. versionadded:: 2012.1(Essex)

Default: ``"127.0.0.1"``

The hostname of the Keystone server used for authentication if you only have
one region. This is often the **only** setting that needs to be set for a
basic deployment.

If you have multiple regions you should use the `AVAILABLE_REGIONS`_ setting
instead.

OPENRC_CUSTOM_TEMPLATE
----------------------

.. versionadded:: 15.0.0(Stein)

Default: ``None``

Example:: ``my-openrc.sh.template``

A template name for a custom user's ``openrc`` file.
``None`` means the default template for ``openrc`` is used.

If the default template is not suitable for your deployment,
for example, if your deployment uses saml2, openid and so on
for authentication, the default ``openrc`` would not be sufficient.
You can provide your own clouds.yaml by specifying this setting.

The default template is defined as :opendev-repo:`openrc.sh.template
<openstack/horizon/src/branch/master/openstack_dashboard/dashboards/project/api_access/templates/api_access/openrc.sh.template>`
and available context parameters are found in ``_get_openrc_credentials()``
and ``download_rc_file()`` functions in
:opendev-repo:`openstack_dashboard/dashboards/project/api_access/views.py
<openstack/horizon/src/branch/master/openstack_dashboard/dashboards/project/api_access/views.py>`

.. note::

   Your template needs to be placed in the search paths of Django templates.
   Check ``TEMPLATES[0]['DIRS']``.
   You may need to specify somewhere your template exist
   to ``DIRS`` in ``TEMPLATES`` setting.

OPENSTACK_PROFILER
------------------

.. versionadded:: 11.0.0(Ocata)

Default: ``{"enabled": False}``

Various settings related to integration with osprofiler library. Since it is a
developer feature, it starts as disabled. To enable it, more than a single
``"enabled"`` key should be specified. Additional keys that should be specified
in that dictionary are:

* ``"keys"`` is a list of strings, which are secret keys used to encode/decode
  the profiler data contained in request headers. Encryption is used for
  security purposes, other OpenStack components that are expected to profile
  themselves with osprofiler using the data from the request that Horizon
  initiated must share a common set of keys with the ones in Horizon
  config. List of keys is used so that security keys could be changed in
  non-obtrusive manner for every component in the cloud.
  Example: ``"keys": ["SECRET_KEY", "MORE_SECRET_KEY"]``.
  For more details see `osprofiler documentation`_.
* ``"notifier_connection_string"`` is a url to which trace messages are sent by
  Horizon. For other components it is usually the only URL specified in config,
  because other components act mostly as traces producers. Example:
  ``"notifier_connection_string": "mongodb://%s" % OPENSTACK_HOST``.
* ``"receiver_connection_string"`` is a url from which traces are retrieved by
  Horizon, needed because Horizon is not only the traces producer, but also a
  consumer. Having 2 settings which usually contain the same value is legacy
  feature from older versions of osprofiler when OpenStack components could use
  oslo.messaging for notifications and the trace client used ceilometer as a
  receiver backend. By default Horizon uses the same URL pointing to a MongoDB
  cluster for both purposes.
  Example: ``"receiver_connection_string": "mongodb://%s" % OPENSTACK_HOST``.

.. _osprofiler documentation: https://docs.openstack.org/osprofiler/latest/user/integration.html#how-to-initialize-profiler-to-get-one-trace-across-all-services

OPENSTACK_SSL_CACERT
--------------------

.. versionadded:: 2013.2(Havana)

Default: ``None``

When unset or set to ``None`` the default CA certificate on the system is used
for SSL verification.

When set with the path to a custom CA certificate file, this overrides use of
the default system CA certificate. This custom certificate is used to verify all
connections to openstack services when making API calls.

OPENSTACK_SSL_NO_VERIFY
-----------------------

.. versionadded:: 2012.2(Folsom)

Default: ``False``

Disable SSL certificate checks in the OpenStack clients (useful for self-signed
certificates).

OPERATION_LOG_ENABLED
---------------------

.. versionadded:: 10.0.0(Newton)

Default: ``False``

This setting can be used to enable logging of all operations carried out by
users of Horizon. The format of the logs is configured via
`OPERATION_LOG_OPTIONS`_

.. note::

  If you use this feature, you need to configure the logger setting like
  an outputting path for operation log in ``local_settings.py``.

OPERATION_LOG_OPTIONS
---------------------

.. versionadded:: 10.0.0(Newton)

.. versionchanged:: 12.0.0(Pike)

    Added ``ignored_urls`` parameter and added ``%(client_ip)s`` to ``format``

Default:

.. code-block:: python

    {
        'mask_fields': ['password'],
        'target_methods': ['POST'],
        'ignored_urls': ['/js/', '/static/', '^/api/'],
        'format': ("[%(domain_name)s] [%(domain_id)s] [%(project_name)s]"
            " [%(project_id)s] [%(user_name)s] [%(user_id)s] [%(request_scheme)s]"
            " [%(referer_url)s] [%(request_url)s] [%(message)s] [%(method)s]"
            " [%(http_status)s] [%(param)s]"),
    }

This setting controls the behavior of the operation log.

* ``mask_fields`` is a list of keys of post data which should be masked from the
  point of view of security. Fields like ``password`` should be included.
  The fields specified in ``mask_fields`` are logged as ``********``.
* ``target_methods`` is a request method which is logged to an operation log.
  The valid methods are ``POST``, ``GET``, ``PUT``, ``DELETE``.
* ``ignored_urls`` is a list of request URLs to be hidden from a log.
* ``format`` defines the operation log format.
  Currently you can use the following keywords.
  The default value contains all keywords.

  * ``%(client_ip)s``
  * ``%(domain_name)s``
  * ``%(domain_id)s``
  * ``%(project_name)s``
  * ``%(project_id)s``
  * ``%(user_name)s``
  * ``%(user_id)s``
  * ``%(request_scheme)s``
  * ``%(referer_url)s``
  * ``%(request_url)s``
  * ``%(message)s``
  * ``%(method)s``
  * ``%(http_status)s``
  * ``%(param)s``

OVERVIEW_DAYS_RANGE
-------------------

.. versionadded:: 10.0.0(Newton)

Default:: ``1``

When set to an integer N (as by default), the start date in the Overview panel
meters will be today minus N days. This setting is used to limit the amount of
data fetched by default when rendering the Overview panel. If set to ``None``
(which corresponds to the behavior in past Horizon versions), the start date
will be from the beginning of the current month until the current date. The

openstack_dashboard/settings.py
# Copyright 2012 United States Government as represented by the
# Administrator of the National Aeronautics and Space Administration.
# All Rights Reserved.
#
# Copyright 2012 Nebula, Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

import glob
import logging
import os
import sys
import warnings

from django.utils.translation import gettext_lazy as _

from horizon.utils.escape import monkeypatch_escape

from openstack_dashboard import enabled
from openstack_dashboard import exceptions
from openstack_dashboard.local import enabled as local_enabled
from openstack_dashboard import theme_settings
from openstack_dashboard.utils import config
from openstack_dashboard.utils import settings as settings_utils

monkeypatch_escape()

# Load default values
# pylint: disable=wrong-import-position
from openstack_dashboard.defaults import *  # noqa: E402,F403,H303

_LOG = logging.getLogger(__name__)

warnings.filterwarnings("default", category=DeprecationWarning)

ROOT_PATH = os.path.dirname(os.path.abspath(__file__))

if ROOT_PATH not in sys.path:
    sys.path.append(ROOT_PATH)

DEBUG = False

ROOT_URLCONF = 'openstack_dashboard.urls'

HORIZON_CONFIG = {
    'user_home': 'openstack_dashboard.views.get_user_home',
    'ajax_queue_limit': 10,
    'auto_fade_alerts': {
        'delay': 3000,
        'fade_duration': 1500,
        'types': ['alert-success', 'alert-info']
    },
    'bug_url': None,
    'help_url': "https://docs.openstack.org/",
    'exceptions': {'recoverable': exceptions.RECOVERABLE,
                   'not_found': exceptions.NOT_FOUND,
                   'unauthorized': exceptions.UNAUTHORIZED},
    'modal_backdrop': 'static',
    'angular_modules': [],
    'js_files': [],
    'js_spec_files': [],
    'external_templates': [],
    'plugins': [],
    'integration_tests_support': INTEGRATION_TESTS_SUPPORT
}

MIDDLEWARE = (
    'openstack_auth.middleware.OpenstackAuthMonkeyPatchMiddleware',
    'debreach.middleware.RandomCommentMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'horizon.middleware.OperationLogMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'horizon.middleware.HorizonMiddleware',
    'horizon.themes.ThemeMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'openstack_dashboard.contrib.developer.profiler.middleware.'
    'ProfilerClientMiddleware',
    'openstack_dashboard.contrib.developer.profiler.middleware.'
    'ProfilerMiddleware',
)

CACHED_TEMPLATE_LOADERS = [
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    'horizon.loaders.TemplateLoader'
]

ADD_TEMPLATE_LOADERS = []
ADD_TEMPLATE_DIRS = []

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(ROOT_PATH, 'templates')],
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.i18n',
                'django.template.context_processors.request',
                'django.template.context_processors.media',
                'django.template.context_processors.static',
                'django.contrib.messages.context_processors.messages',
                'horizon.context_processors.horizon',
                'openstack_dashboard.context_processors.openstack',
            ],
            'loaders': [
                'horizon.themes.ThemeTemplateLoader'
            ],
        },
    },
]

STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'horizon.contrib.staticfiles.finders.HorizonStaticFinder',
    'compressor.finders.CompressorFinder',
)

COMPRESS_PRECOMPILERS = (
    ('text/scss', 'horizon.utils.scss_filter.HorizonScssFilter'),
)

COMPRESS_CSS_FILTERS = (
    'compressor.filters.css_default.CssAbsoluteFilter',
)

COMPRESS_ENABLED = True
COMPRESS_OUTPUT_DIR = 'dashboard'
COMPRESS_CSS_HASHING_METHOD = 'hash'
COMPRESS_PARSER = 'compressor.parser.HtmlParser'

INSTALLED_APPS = [
    'openstack_dashboard',
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.humanize',
    'openstack_dashboard.django_pyscss_fix',
    'django_pyscss',
    'debreach',
    'compressor',
    'horizon',
    'openstack_auth',
]

AUTHENTICATION_BACKENDS = ('openstack_auth.backend.KeystoneBackend',)
AUTH_USER_MODEL = 'openstack_auth.User'
MESSAGE_STORAGE = 'django.contrib.messages.storage.fallback.FallbackStorage'

SESSION_ENGINE = 'django.contrib.sessions.backends.cache'
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
        'LOCATION': '127.0.0.1:11211',
    },
}

SESSION_COOKIE_HTTPONLY = True
SESSION_EXPIRE_AT_BROWSER_CLOSE = True
SESSION_COOKIE_SECURE = False

# when doing upgrades, it may be wise to stick to PickleSerializer
# NOTE(berendt): Check during the K-cycle if this variable can be removed.
#                https://bugs.launchpad.net/horizon/+bug/1349463
SESSION_SERIALIZER = 'django.contrib.sessions.serializers.PickleSerializer'

CSRF_FAILURE_VIEW = 'openstack_dashboard.views.csrf_failure'

LANGUAGES = (
    ('cs', 'Czech'),
    ('de', 'German'),
    ('en', 'English'),
    ('en-au', 'Australian English'),
    ('en-gb', 'British English'),
    ('eo', 'Esperanto'),
    ('es', 'Spanish'),
    ('fr', 'French'),
    ('id', 'Indonesian'),
    ('it', 'Italian'),
    ('ja', 'Japanese'),
    ('ko', 'Korean (Korea)'),
    ('pl', 'Polish'),
    ('pt-br', 'Portuguese (Brazil)'),
    ('ru', 'Russian'),
    ('tr', 'Turkish'),
    ('zh-hans', 'Simplified Chinese'),
    ('zh-hant', 'Traditional Chinese'),
)
LANGUAGE_CODE = 'en'
LANGUAGE_COOKIE_NAME = 'horizon_language'
USE_I18N = True
USE_L10N = True
USE_TZ = True

DEFAULT_EXCEPTION_REPORTER_FILTER = 'horizon.exceptions.HorizonReporterFilter'

SECRET_KEY = None
LOCAL_PATH = None

ADD_INSTALLED_APPS = []

CSRF_COOKIE_AGE = None

COMPRESS_OFFLINE_CONTEXT = 'horizon.themes.offline_context'

# Notice all customizable configurations should be above this line
XSTATIC_MODULES = settings_utils.BASE_XSTATIC_MODULES

OPENSTACK_PROFILER = {
    'enabled': False
}

CHAMELEON_PORTAL_API_BASE_URL = 'https://www.chameleoncloud.org'
APPLIANCE_CATALOG_API_PATH = '/appliances/api/appliances'
APPLIANCE_CATALOG_DETAILS_PATH = '/appliances'
PUBLISH_APPLIANCE_PATH = '/appliances/create/image'

if not LOCAL_PATH:
    LOCAL_PATH = os.path.join(ROOT_PATH, 'local')
LOCAL_SETTINGS_DIR_PATH = os.path.join(LOCAL_PATH, "local_settings.d")

_files = glob.glob(os.path.join(LOCAL_PATH, 'local_settings.conf'))
_files.extend(
    sorted(glob.glob(os.path.join(LOCAL_SETTINGS_DIR_PATH, '*.conf'))))
_config = config.load_config(_files, ROOT_PATH, LOCAL_PATH)

# Apply the general configuration.
config.apply_config(_config, globals())

try:
    from local.local_settings import *  # noqa: F403,H303
except ImportError:
    _LOG.warning("No local_settings file found.")

# configure templates
if not TEMPLATES[0]['DIRS']:
    TEMPLATES[0]['DIRS'] = [os.path.join(ROOT_PATH, 'templates')]

TEMPLATES[0]['DIRS'] += ADD_TEMPLATE_DIRS

# configure template debugging
TEMPLATES[0]['OPTIONS']['debug'] = DEBUG

# Template loaders
if DEBUG:
    TEMPLATES[0]['OPTIONS']['loaders'].extend(
        CACHED_TEMPLATE_LOADERS + ADD_TEMPLATE_LOADERS
    )
else:
    TEMPLATES[0]['OPTIONS']['loaders'].extend(
        [('django.template.loaders.cached.Loader', CACHED_TEMPLATE_LOADERS)] +
        ADD_TEMPLATE_LOADERS
    )

# allow to drop settings snippets into a local_settings_dir
LOCAL_SETTINGS_DIR_PATH = os.path.join(ROOT_PATH, "local", "local_settings.d")
if os.path.exists(LOCAL_SETTINGS_DIR_PATH):
    for (dirpath, dirnames, filenames) in os.walk(LOCAL_SETTINGS_DIR_PATH):
        for filename in sorted(filenames):
            if filename.endswith(".py"):
                try:
                    with open(os.path.join(dirpath, filename)) as f:
                        # pylint: disable=exec-used
                        exec(f.read())
                except Exception:
                    _LOG.exception(
                        "Can not exec settings snippet %s", filename)

if USER_MENU_LINKS is None:
    USER_MENU_LINKS = []
    if SHOW_OPENRC_FILE:
        USER_MENU_LINKS.append({
            'name': _('OpenStack RC File'),
            'icon_classes': ['fa-download', ],
            'url': 'horizon:project:api_access:openrc',
        })

if not WEBROOT.endswith('/'):
    WEBROOT += '/'
if LOGIN_URL is None:
    LOGIN_URL = WEBROOT + 'auth/login/'
if LOGOUT_URL is None:
    LOGOUT_URL = WEBROOT + 'auth/logout/'
if LOGIN_ERROR is None:
    LOGIN_ERROR = WEBROOT + 'auth/error/'
if LOGIN_REDIRECT_URL is None:
    LOGIN_REDIRECT_URL = WEBROOT

if MEDIA_ROOT is None:
    MEDIA_ROOT = os.path.abspath(os.path.join(ROOT_PATH, '..', 'media'))

if MEDIA_URL is None:
    MEDIA_URL = WEBROOT + 'media/'

if STATIC_ROOT is None:
    STATIC_ROOT = os.path.abspath(os.path.join(ROOT_PATH, '..', 'static'))

if STATIC_URL is None:
    STATIC_URL = WEBROOT + 'static/'

AVAILABLE_THEMES, SELECTABLE_THEMES, DEFAULT_THEME = (
    theme_settings.get_available_themes(
        AVAILABLE_THEMES,
        DEFAULT_THEME,
        SELECTABLE_THEMES
    )
)

# Discover all the directories that contain static files
STATICFILES_DIRS = theme_settings.get_theme_static_dirs(
    AVAILABLE_THEMES, THEME_COLLECTION_DIR, ROOT_PATH)

# Ensure that we always have a SECRET_KEY set, even when no local_settings.py
# file is present. See local_settings.py.example for full documentation on the
# horizon.utils.secret_key module and its use.
if not SECRET_KEY:
    if not LOCAL_PATH:
        LOCAL_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                  'local')

    # pylint: disable=ungrouped-imports
    from horizon.utils import secret_key
    SECRET_KEY = secret_key.generate_or_read_from_file(os.path.join(LOCAL_PATH,
                                                       '.secret_key_store'))

# populate HORIZON_CONFIG with auto-discovered JavaScript sources, mock files,
# specs files and external templates.
settings_utils.find_static_files(HORIZON_CONFIG, AVAILABLE_THEMES,
                                 THEME_COLLECTION_DIR, ROOT_PATH)

INSTALLED_APPS = list(INSTALLED_APPS)  # Make sure it's mutable
settings_utils.update_dashboards(
    [
        enabled,
        local_enabled,
    ],
    HORIZON_CONFIG,
    INSTALLED_APPS,
)
INSTALLED_APPS[0:0] = ADD_INSTALLED_APPS

# Include xstatic_modules specified in plugin
XSTATIC_MODULES += HORIZON_CONFIG['xstatic_modules']

# Discover all the xstatic module entry points to embed in our HTML
STATICFILES_DIRS += settings_utils.get_xstatic_dirs(
    XSTATIC_MODULES, HORIZON_CONFIG)

# This base context objects gets added to the offline context generator
# for each theme configured.
HORIZON_COMPRESS_OFFLINE_CONTEXT_BASE = {
    'WEBROOT': WEBROOT,
    'STATIC_URL': STATIC_URL,
    'HORIZON_CONFIG': HORIZON_CONFIG,
    'NG_TEMPLATE_CACHE_AGE': NG_TEMPLATE_CACHE_AGE if not DEBUG else 0,
}

if DEBUG:
    logging.basicConfig(level=logging.DEBUG)


# Here comes the Django settings deprecation section. Being at the very end
# of settings.py allows it to catch the settings defined in local_settings.py
# or inside one of local_settings.d/ snippets.
